src/mscorlib/shared/System/Memory.cs

```diff
@@ -0,0 +1,287 @@
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public Memory(T[] array)
+        {
+            if (array == null)
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.array);
+            if (default(T) == null && array.GetType() != typeof(T[]))
+                ThrowHelper.ThrowArrayTypeMismatchException();
+
+            _arrayOrOwnedMemory = array;
+            _index = 0;
+            _length = array.Length;
+        }
+
+        /// <summary>
+        /// Creates a new memory over the portion of the target array beginning
+        /// at 'start' index and ending at 'end' index (exclusive).
+        /// </summary>
+        /// <param name="array">The target array.</param>
+        /// <param name="start">The index at which to begin the memory.</param>
+        /// <param name="length">The number of items in the memory.</param>
+        /// <exception cref="System.ArgumentNullException">Thrown when <paramref name="array"/> is a null
+        /// reference (Nothing in Visual Basic).</exception>
+        /// <exception cref="System.ArrayTypeMismatchException">Thrown when <paramref name="array"/> is covariant and array's type is not exactly T[].</exception>
+        /// <exception cref="System.ArgumentOutOfRangeException">
+        /// Thrown when the specified <paramref name="start"/> or end index is not in the range (&lt;0 or &gt;=Length).
+        /// </exception>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public Memory(T[] array, int start, int length)
+        {
+            if (array == null)
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.array);
+            if (default(T) == null && array.GetType() != typeof(T[]))
+                ThrowHelper.ThrowArrayTypeMismatchException();
+            if ((uint)start > (uint)array.Length || (uint)length > (uint)(array.Length - start))
+                ThrowHelper.ThrowArgumentOutOfRangeException();
+
+            _arrayOrOwnedMemory = array;
+            _index = start;
+            _length = length;
+        }
+        
+        // Constructor for internal use only.
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal Memory(OwnedMemory<T> owner, int index, int length)
+        {
+            if (owner == null)
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.ownedMemory);
+            if (index < 0 || length < 0)
+                ThrowHelper.ThrowArgumentOutOfRangeException();
+
+            _arrayOrOwnedMemory = owner;
+            _index = index | (1 << 31); // Before using _index, check if _index < 0, then 'and' it with RemoveOwnedFlagBitMask
+            _length = length;
+        }
+
+        /// <summary>
+        /// Defines an implicit conversion of an array to a <see cref="Memory{T}"/>
+        /// </summary>
+        public static implicit operator Memory<T>(T[] array) => new Memory<T>(array);
```

Comment:
❓ Since this is an `implicit` conversion, why are we not handling `null` by returning `default`?

```csharp
=> array != null ? new Memory<T>(array) : default;
```

Reply:
I am following the same semantics as Span:
https://github.com/dotnet/coreclr/blob/master/src/mscorlib/shared/System/Span.cs#L325

Reply:
❓ Is `Memory<T>` supposed to be a `ref struct` in the end?

Reply:
No. It is what Span is for. Memory is non-ref counter-part of Span.

Memory design doc: https://github.com/dotnet/corefxlab/blob/master/docs/specs/memory.md